test/extensions/filters/http/geoip/geoip_filter_test.cc

```diff
@@ -0,0 +1,278 @@
+                                                    {"x-geo-anon-proxy", "true"}};
+  for (auto iter = geo_headers.begin(); iter != geo_headers.end(); ++iter) {
+    auto& header = iter->first;
+    expectStats(header);
+  }
+  Network::Address::InstanceConstSharedPtr remote_address =
+      Network::Utility::parseInternetAddress("1.2.3.4");
+  filter_callbacks_.stream_info_.downstream_connection_info_provider_->setRemoteAddress(
+      remote_address);
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_->decodeHeaders(request_headers, false));
+  EXPECT_CALL(filter_callbacks_, continueDecoding());
+  filter_->onLookupComplete(dummy_region_, "x-geo-region");
+  filter_->onLookupComplete(dummy_country_, "x-geo-country");
+  filter_->onLookupComplete(dummy_city_, "x-geo-city");
+  filter_->onLookupComplete(dummy_asn_, "x-geo-asn");
+  filter_->onLookupComplete(dummy_anon_response_, "x-geo-anon");
+  filter_->onLookupComplete(dummy_anon_response_, "x-geo-anon-vpn");
+  filter_->onLookupComplete(dummy_anon_response_, "x-geo-anon-hosting");
+  filter_->onLookupComplete(dummy_anon_response_, "x-geo-anon-tor");
+  filter_->onLookupComplete(dummy_anon_response_, "x-geo-anon-proxy");
+  EXPECT_EQ(9, request_headers.size());
+  for (auto iter = geo_headers.begin(); iter != geo_headers.end(); ++iter) {
+    auto& header = iter->first;
+    auto& value = iter->second;
+    expectHeader(request_headers, header, value);
+  }
+  EXPECT_EQ("1.2.3.4:0", captured_address_->asString());
+}
+
+TEST_F(GeoipFilterTest, GeoHeaderNotAppendedOnEmptyLookup) {
+  initializeProviderFactory();
+  const std::string external_request_yaml = R"EOF(
+    geo_headers_to_add:
+      region: "x-geo-region"
+      city:   "x-geo-city"
+    provider:
+        name: "envoy.geoip_providers.dummy"
+)EOF";
+  initializeFilter(external_request_yaml);
+  Http::TestRequestHeaderMapImpl request_headers;
+  EXPECT_CALL(stats_, counter("prefix.geoip.x-geo-city.total"));
+  EXPECT_CALL(stats_, counter("prefix.geoip.x-geo-city.hit")).Times(0);
+  EXPECT_CALL(stats_, counter("prefix.geoip.x-geo-region.total"));
+  EXPECT_CALL(stats_, counter("prefix.geoip.x-geo-region.hit"));
+  Network::Address::InstanceConstSharedPtr remote_address =
+      Network::Utility::parseInternetAddress("1.2.3.4");
+  filter_callbacks_.stream_info_.downstream_connection_info_provider_->setRemoteAddress(
+      remote_address);
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_->decodeHeaders(request_headers, false));
+  EXPECT_CALL(filter_callbacks_, continueDecoding());
+  filter_->onLookupComplete(dummy_region_, "x-geo-region");
+  filter_->onLookupComplete(empty_response_, "x-geo-city");
+
+  EXPECT_EQ(1, request_headers.size());
+  EXPECT_EQ("dummy_region",
+            request_headers.get(Http::LowerCaseString("x-geo-region"))[0]->value().getStringView());
+  EXPECT_EQ("1.2.3.4:0", captured_address_->asString());
+}
+
```

Comment:
Combination of the other comments - please add a test `NoCrashIfFilterDestroyedBeforeCallbackCalled`, in which filter_->decodeHeaders is called and queues up some driver callbacks, then `filter_.clear()`, then call the callbacks.